model/riscv_insts_zbkb.sail

```diff
@@ -0,0 +1,104 @@
+  let rs1_val = X(rs1);
+  let rs2_val = X(rs2);
+  let result : xlenbits = match op {
+    RISCV_PACK  => if sizeof(xlen) == 32
+                   then rs2_val[15..0] @ rs1_val[15..0]
+                   else rs2_val[31..0] @ rs1_val[31..0],
+    RISCV_PACKH => EXTZ(rs2_val[7..0] @ rs1_val[7..0])
+  };
+  X(rd) = result;
+  RETIRE_SUCCESS
+}
+
+/* ****************************************************************** */
+union clause ast = ZBKB_PACKW : (regidx, regidx, regidx)
+
+mapping clause encdec = ZBKB_PACKW(rs2, rs1, rd) if haveZbkb() & sizeof(xlen) == 64
+  <-> 0b0000100 @ rs2 @ rs1 @ 0b100 @ rd @ 0b0111011 if haveZbkb() & sizeof(xlen) == 64
+
+mapping clause assembly = ZBKB_PACKW(rs2, rs1, rd)
+  <-> "packw" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
+
+function clause execute (ZBKB_PACKW(rs2, rs1, rd)) = {
+  let rs1_val = X(rs1);
+  let rs2_val = X(rs2);
+  let result : bits(32) = rs2_val[15..0] @ rs1_val[15..0];
+  X(rd) = EXTZ(result);
+  RETIRE_SUCCESS
+}
+
+/* ****************************************************************** */
+union clause ast = RISCV_ZIP : (regidx, regidx)
+
+mapping clause encdec = RISCV_ZIP(rs1, rd) if haveZbkb() & sizeof(xlen) == 32
+  <-> 0b000010001111 @ rs1 @ 0b001 @ rd @ 0b0010011 if haveZbkb() & sizeof(xlen) == 32
+
+mapping clause assembly = RISCV_ZIP(rs1, rd)
+  <-> "zip" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+
+function clause execute (RISCV_ZIP(rs1, rd)) = {
+  let rs1_val = X(rs1);
+  result : xlenbits = zeros();
+  foreach (i from 0 to 15) {
+    result[i*2] = rs1_val[i];
+    result[i*2 + 1] = rs1_val[i + 16];
+  };
+  X(rd) = result;
+  RETIRE_SUCCESS
+}
+
+/* ****************************************************************** */
+union clause ast = RISCV_UNZIP : (regidx, regidx)
+
+mapping clause encdec = RISCV_UNZIP(rs1, rd) if haveZbkb() & sizeof(xlen) == 32
+  <-> 0b000010001111 @ rs1 @ 0b101 @ rd @ 0b0010011 if haveZbkb() & sizeof(xlen) == 32
+
+mapping clause assembly = RISCV_UNZIP(rs1, rd)
+  <-> "unzip" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+
+function clause execute (RISCV_UNZIP(rs1, rd)) = {
+  let rs1_val = X(rs1);
+  result : xlenbits = zeros();
+  foreach (i from 0 to 15) {
+    result[i]      = rs1_val[i*2];
+    result[i + 16] = rs1_val[i*2 + 1];
+  };
+  X(rd) = result;
+  RETIRE_SUCCESS
+}
+/* ****************************************************************** */
+union clause ast = RISCV_BREV8 : (regidx, regidx)
+
+mapping clause encdec = RISCV_BREV8(rs1, rd) if haveZbkb()
+  <-> 0b011010000111 @ rs1 @ 0b101 @ rd @ 0b0010011 if haveZbkb()
+
+mapping clause assembly = RISCV_BREV8(rs1, rd)
+  <-> "brev8" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+
+function clause execute (RISCV_BREV8(rs1, rd)) = {
+  let rs1_val = X(rs1);
+  result : xlenbits = zeros();
+  foreach (i from 0 to (sizeof(xlen) - 8) by 8)
+    foreach (j from i to (i + 7))
+      result[j] = rs1_val[(i + 7) - j];
```

Comment:
Doesn't this produce 4 or 8 (depending on XLEN) copies of the reversed rs1_val[7..0]?

This also would probably be better by defining a `val bitvector_reverse : forall 'n. bits('n) -> bits('n)` and `overload reverse = {bitvector_reverse}` (split that way in case anyone wants to add a vector_reverse for something else; it's done like that in the standard library) that you can then use here rather than needing an inner loop.

Reply:
Yes it does produce 4 or 8 (depending on XLEN) copies of the reversed rs1_val[7..0].

Sure, I'll update by using bitvector_reverse.

Reply:
To be clear: this should *not* make multiple copies of the same byte; 
it should be reversing each byte individually, e.g.: 
  foreach (j from 0 to  7)
      result[i + j] = rs1_val[i + (7 - j)];
but using a bitvector reverse  function is far more readable, and can be used elsewhere, as pointed out,  in  which case
   foreach (i from 0 to (sizeof(xlen) - 8) by 8)
     result[i..i + 7] = bitreverse[rs1_val[i..i + 7];
(I won't pretend this syntax is exactly correct]

Reply:
Yep, exactly

Reply:
@allenjbaum The function I wrote with nested for loops is working perfectly, reversing each byte individually. Yeah but I'll implement bitvector reverse function as suggested by Jessica. 
@jrtc27 A quick question, should I add bitvector reverse function in preludes.sail or riscv_insts_zbkb.sail or it doesn't make any difference where I add the function (overloading)?

Reply:
Prelude, it's a general utility function that should exist in the standard library but doesn't